Support Route.spec.subdomain

jacksgt opened the pull request:
Hi there!

As discussed at KubeCon, I had this patch sitting on my machine for quite a while - sorry for the delay!

In summary, this patch implements support for the OpenShift Route "subdomain" enhancement (added in OpenShift 4.11).
For reference, this is the commit that introduces the change: https://github.com/openshift/router/commit/6f730c7cae966f0ed8def50c81d1bf10fe9eb77b

And here are the release notes for 4.11: https://docs.openshift.com/container-platform/4.11/release_notes/ocp-4-11-release-notes.html#ocp-4-11-routes-spec-subdomain-field

> Previously, you could not specify the subdomain of a route, and the spec.host field was required to set the host name. You can now specify the spec.subdomain field and omit the spec.host field of a route. The router deployment that exposes the route will use the spec.subdomain value to determine the host name.

Closes https://github.com/cert-manager/openshift-routes/issues/10

----

Status of this PR:

The basic functionality should work, but I haven't added any tests yet because I would first like to know if the maintainers are comfortable with the direction of the code. :-)
I will add tests once we have an agreement.

maelvls commented:
Hi! Thank you for submitting this PR! I'll take a look shortly (hopefully this afternoon).

maelvls commented on internal/controller/sync.go:
```diff
@@ -294,15 +298,22 @@ func (r *Route) createNextCR(ctx context.Context, route *routev1.Route, revision
 	var dnsNames []string
+	// Get the canonical hostname(s) of the Route (from .spec.host or .spec.subdomain)
+	dnsNames = getRouteHostnames(route)
+	if len(dnsNames) <= 0 {
```
Are negative integers possible values returned by `getRouteHostnames`? If not, I'd avoid the ambiguity created by `<=`:

```suggestion
	if len(dnsNames) == 0 {
```

maelvls commented on internal/controller/sync.go:
```diff
@@ -487,3 +498,34 @@ func certDurationFromRoute(r *routev1.Route) (time.Duration, error) {
 	return duration, nil
 }
+
+// This function returns the hostnames that have been admitted by an Ingress Controller
```
Tiny nitpick 😇
```suggestion
// This function returns the hostnames that have been admitted by an Ingress Controller.
```

maelvls commented on internal/controller/sync.go:
```diff
@@ -487,3 +498,34 @@ func certDurationFromRoute(r *routev1.Route) (time.Duration, error) {
+// Usually this is just `.spec.host`, but as of OpenShift 4.11 users may also specify `.spec.subdomain`,
+// in which case the fully qualified hostname is derived from the hostname of the Ingress Controller.
+// In both cases, the final hostname is reflected in `.status.ingress[].host`.
+func getRouteHostnames(r *routev1.Route) []string {
```
Do we really need to return a list of domains here? It seems like one Route = one host

jacksgt replied:
Yes we need, because a Route can be admitted by multiple ingress controllers. I had already prepared this example but forgot to attach it to this PR - my bad!

```yaml
apiVersion: route.openshift.io/v1
kind: Route
metadata:
  creationTimestamp: "2023-05-03T07:10:21Z"
  name: test-subdomain
  namespace: openshift-cern-ingress-health-checks
  resourceVersion: "860465"
  uid: b6f0e564-ce65-43c8-8bb4-d989e6d76ed4
spec:
  subdomain: test-subdomain
  tls:
    termination: edge
  to:
    kind: Service
    name: health-checks-apps-shard-1
    weight: 100
  wildcardPolicy: None
status:
  ingress:
  - conditions:
    - lastTransitionTime: "2023-05-03T07:10:21Z"
      status: "True"
      type: Admitted
    host: test-subdomain.okd-apps-2.example.com
    routerCanonicalHostname: router-okd-apps-2.okd-apps-2.example.com
    routerName: okd-apps-2
    wildcardPolicy: None
  - conditions:
    - lastTransitionTime: "2023-05-03T07:10:21Z"
      status: "True"
      type: Admitted
    host: test-subdomain.okd-apps-1.example.com
    routerCanonicalHostname: router-okd-apps-1.okd-apps-1.example.com
    routerName: router-okd-apps-1
    wildcardPolicy: None
  - conditions:
    - lastTransitionTime: "2023-05-03T07:10:21Z"
      status: "True"
      type: Admitted
    host: test-subdomain.okd.example.com
    routerCanonicalHostname: router-default.okd.example.com
    routerName: default
    wildcardPolicy: None
```

This route is available at `test-subdomain.okd.example.com` as well as `test-subdomain.okd-apps-1.example.com`.

maelvls replied:
Thanks, that totally makes sense.

maelvls reviewed (commented):
Hey, these changes look great. Thanks for the well-written comments in code, that helped a lot.

~I wasn't able to test yet because the image `openshift/origin-haproxy-router:v4.11.0` has not been published yet. Also, I don't know how to build that image, the [upstream Dockerfile](https://github.com/openshift/router/blob/master/images/router/haproxy/Dockerfile) refers to a registry for which I don't have access 😬~ The image is available at: quay.io/openshift/origin-haproxy-router:4.11.

For reference, here are the steps I wanted to take to test the feature. I wish we had some automated tests, but I don't really have the time to write some at the moment:

1. Install OpenShift Route CRDs and controller:

    ```bash
    kind create cluster
    kubectl apply -f https://raw.githubusercontent.com/openshift/router/master/deploy/route_crd.yaml
    kubectl apply -f https://raw.githubusercontent.com/openshift/router/master/deploy/router_rbac.yaml
    kubectl apply -f https://raw.githubusercontent.com/openshift/router/master/deploy/router.yaml
    kubectl set image -n openshift-ingress deploy/ingress-router router=quay.io/openshift/origin-haproxy-router:4.11
    kubectl set env -n openshift-ingress deploy/ingress-router ROUTER_DOMAIN=domain.com
    kubectl apply -f https://github.com/jetstack/cert-manager/releases/download/v1.13.0/cert-manager.yaml
    ```

2. Run openshift-routes:
   ```bash
   go run ./internal/cmd/ -v=5
   ```

3. Create a cert-manager ClusterIssuer:

    ```bash
    kubectl apply -f- <<EOF
    apiVersion: cert-manager.io/v1
    kind: Issuer
    metadata:
      name: self-signed
      namespace: cert-manager
    spec:
      selfSigned: {}
    ---
    apiVersion: cert-manager.io/v1
    kind: Certificate
    metadata:
      name: example
      namespace: cert-manager
    spec:
      isCA: true
      privateKey:
        algorithm: ECDSA
        size: 256
      secretName: example
      commonName: example root
      duration: 262800h
      issuerRef:
        name: self-signed
        kind: Issuer
    ---
    apiVersion: cert-manager.io/v1
    kind: ClusterIssuer
    metadata:
      name: example
      namespace: cert-manager
    spec:
      ca:
        secretName: example
    EOF
    ```

4. Create a Route with the subdomain `foo`. Since openshift-ingress has been
   configured with `ROUTER_DOMAIN=domain.com`, we don't need to specify the full
   domain. The actual hostname will be `foo.domain.com`.

    ```bash
    kubectl apply -f- <<EOF
    apiVersion: route.openshift.io/v1
    kind: Route
    metadata:
      name: example-route
      annotations:
        cert-manager.io/issuer-name: example
        cert-manager.io/issuer-kind: ClusterIssuer
    spec:
      subdomain: foo
      wildcardPolicy: None
      to:
        name: httpbin
    EOF
    ```

5. Get the IP of the node (WARNING: this part only works from the Docker host; it likely won't work on macOS):

    ```bash
    IP=$(kubectl get nodes -ojson | jq '.items[].status.addresses[] | select(.type == "InternalIP").address' -r)
    ```

6. Confirm that the issued cert's SAN is `foo.domain.com`:

    ```console
    openssl s_client -connect $IP:443 -servername foo.domain.com 2>/dev/null <<<"" \
      | openssl x509 -noout -text \
      |  awk '/Subject: C=/{printf $NF"\n"} /DNS:/{x=gsub(/ *DNS:/, ""); printf "SANS=" $0"\n"}'
    ```

    It should show:

    ```text
    SANS=foo.domain.com
    ```

maelvls commented:
Hey, great work on this! Really happy to see you around. I haven't had time to review today, but I may be able to review on Friday. I'll keep you updated!

jacksgt commented:
> I was able to reproduce your tests successfully, well done! I confirm that am comfortable with the direction of the code.
> 
> I wish we had some end-to-end tests in place to check this new behaviour but unfortunately we don't have any in place. disappointed I will open an issue to mention the lack of automated tests.
> 
> I think this PR is good to go (unless you thought of other unit tests that I may not have thought of?). Let me know if the PR is good on your side and I'll LGTM.

Thanks for taking another look. Yes, I will add some unit tests for this new feature, then the PR should be good to go.

maelvls commented on internal/controller/sync.go:
```diff
@@ -63,47 +68,58 @@ func (r *Route) sync(ctx context.Context, req reconcile.Request, route *routev1.
+			return result, nil
+		}
+
+		// create CR and return. We own the CR so it will cause a re-reconcile
+		_, err = r.certClient.CertmanagerV1().CertificateRequests(route.Namespace).Create(ctx, cr, metav1.CreateOptions{})
+		if err != nil {
+			return result, err
+		}
+		r.eventRecorder.Event(route, corev1.EventTypeNormal, ReasonIssuing, "Created new CertificateRequest for Route %s")
```
Shouldn't we return right after creating the CR? The CR will trigger a reconciliation of the Route, so we can stop now, no?

jacksgt replied:
You're totally right! This got lost in the refactor. Fixed now.

maelvls reviewed (approved):
Hey, thanks for unit testing `buildNextCR`, that's great work! The refactoring is welcome too. I had not paid attention to the fact that there were some missing unit tests in my previous review. "1337" as the revision number is a nice touch 😅

jacksgt commented:
Hi,
I squashed the commits and added the DCO / Signed-off-by.
I also performed a final validation according to the tests you outlined above.
From my side this is ready.

jacksgt commented:
ping @maelvls

maelvls commented:
Hey, sorry for the delay! I came back from vacation last week but had not got around to review your PR. 😅 Thank you so much for adding the missing unit tests on `buildNextCR` (which was called `createNextCR` previously and wasn't unit tested).

I think it is good to go!

/lgtm

maelvls commented:
Let's wait for #28, #29, and #32 to get merged and let's create a new version of openshift-routes!
